### PR TITLE
[CPP Graph] Baichuan & Baichuan2 Enabling

### PR DESCRIPTION
**Pull request by Zhenzhong1**
## Type of Change

Feature add and Issues Fix

## Description

* [Baichuan-13B-Chat](https://huggingface.co/baichuan-inc/Baichuan-13B-Chat)
* [Baichuan2-13B-Chat](https://huggingface.co/baichuan-inc/Baichuan2-13B-Chat)

baichuan

Add extension-test

  - https://github.com/intel-innersource/frameworks.ai.lpot.lpot-validation/pull/523
  - Baichuan-13B-Chat extension tests https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/148/

## Expected Behavior & Potential Risk

N/A.

## How has this PR been tested?

Manual and CI tests.

## Dependency Change?

N/A.

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp [hunk 1]
```diff
@@ -214,7 +217,8 @@ int main(int argc, char** argv) {
       embd_inp.emplace_back(i);
     }
   } else {
-    embd_inp = ::model_tokenize(ctx, params.prompt, add_bos);
+    //embd_inp = ::model_tokenize(ctx, params.prompt, add_bos);
+    embd_inp.insert(embd_inp.begin(), {195, 9875, 31213, 32889, 196});
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp:221]: draft？

**Comment by airMeng** [comment]
please update KERNEL_DEBUG settings as zhenwei issued during the meeting, and prepare readme

File: intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp (hunk 1 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp]: Yes. Fixed.

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.h [hunk 2]
```diff
@@ -260,6 +260,7 @@ NE_API struct ne_tensor* ne_norm(struct ne_context* ctx, struct ne_tensor* a);
 
 NE_API struct ne_tensor* ne_rms_norm(struct ne_context* ctx, struct ne_tensor* a);
 
+NE_API struct ne_tensor* ne_rms_norm_inplace(struct ne_context* ctx, struct ne_tensor* a);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.h:263]: seems ```rmsnorm``` haven't been supported in jblas epilogue? will we support it? @zhewang1-intc

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+enum baichuan_model {
+  BAICHUAN_UNKNOWN,
+  BAICHUAN_13B,
+};
+
+static const model_scratch baichuan_mem_req(int n_layers) {
+  switch (n_layers) {
+    case 40:
+      return {8192ull * MB, 8192ull * MB, 8192ull * MB};
+    // TODO(hengyu): add more variants besides 6B
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h:30]: ?

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+    // ffn GEMM
+    layer.ffn[0] = ml->get_tensor(layers_i + ".mlp.gate_proj.weight",
+                                  {n_embd, uint32_t(model.hparams.inner_hidden_size)}, backend);
+    layer.ffn[1] =
+        ml->get_tensor(layers_i + ".mlp.up_proj.weight", {n_embd, uint32_t(model.hparams.inner_hidden_size)}, backend);
+    layer.ffn[2] = ml->get_tensor(layers_i + ".mlp.down_proj.weight",
+                                  {uint32_t(model.hparams.inner_hidden_size), n_embd}, backend);
+
+    layer.k_cache = d_ne_new_tensor_3d(model.ctx, NE_TYPE_F16, 5120 / 40, 4096, 40);  // [n_head, maxlen, head_size]
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp:148]: if these are static number, set it as
```suggestion
   static const size_t n_head = 128;
   static const size_t max_len = 4096;
   static const size_t head_size = 40;
   layer.k_cache = d_ne_new_tensor_3d(model.ctx, NE_TYPE_F16, n_head, maxlen, head_size);  // [n_head, maxlen, head_size]
```

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  fprintf(stderr, "%s: n_ff       = %u\n", __func__, n_ff);
+  fprintf(stderr, "%s: n_parts    = %zu\n", __func__, ml->file_loaders.size());
+  n_embd = hparams.n_embd;
+  n_vocab = hparams.n_vocab;
+  n_layer = hparams.n_layer;
+  scratch = baichuan_mem_req(n_layer);
+  model.scratchs = scratch;
+}
+
+#define MODEL_BACKEND_OFFLOAD NE_BACKEND_CPU
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp:85]: useless?

**Review by airMeng — APPROVED** [review]
LGTM except some comments

**Comment by airMeng** [comment]
> please update KERNEL_DEBUG settings as zhenwei issued during the meeting, and prepare readme

better to finish this in this PR

**Comment by Zhenzhong1** [comment]
> > please update KERNEL_DEBUG settings as zhenwei issued during the meeting, and prepare readme
> 
> better to finish this in this PR

I've removed this feature in this PR, because printing data info using a unified API is not something that can be solved with a few simple lines of code, which means it requires more efforts.

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp (hunk 5 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp:85]: All xxx_utils.cpp files have this macro declaration.

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h (hunk 3 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h:30]: All head files of model have this comment and mention you.

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h:30]: quite sad to be notified of this...

Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h:30]: let me remove these commnets in this PR.

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    // ffn GEMM
+    layer.ffn[0] = ml->get_tensor(layers_i + ".mlp.gate_proj.weight",
+                                  {n_embd, uint32_t(model.hparams.inner_hidden_size)}, backend);
+    layer.ffn[1] =
+        ml->get_tensor(layers_i + ".mlp.up_proj.weight", {n_embd, uint32_t(model.hparams.inner_hidden_size)}, backend);
+    layer.ffn[2] = ml->get_tensor(layers_i + ".mlp.down_proj.weight",
+                                  {uint32_t(model.hparams.inner_hidden_size), n_embd}, backend);
+
+    layer.k_cache = d_ne_new_tensor_3d(model.ctx, NE_TYPE_F16, 5120 / 40, 4096, 40);  // [n_head, maxlen, head_size]
+    layer.v_cache = d_ne_new_tensor_3d(model.ctx, NE_TYPE_F16, 4096, 5120 / 40, 40);  // [n_head, head_size, maxlen]
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp:149]: Initialize it as nullptr and kv_cache_init() in model_utils.cpp will initialize kvcache of every layers for you after #435

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.h (hunk 2 above)
Comment (zhewang1-intc) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.h:263]: we have no plan to support rmsnorm in epilogue now, it's hard to calculate rmsnorm in epilogue in current parallel strategy ( we split the dst matrix on N-dim).

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp (hunk 6 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp:149]: I will apply MHA to chatglm and baichuan next week in another PR.

**Comment by a32543254** [comment]
How about FFN  fusion / MHA fusion /

**Comment by Zhenzhong1** [comment]
> Do we enable FFN fusion / MHA fusion in this pr?

No. The MHA & FFN fusion to the chatglm & baichuan will be applied in the next PR. I plan to add models first this week.

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp (hunk 4 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan_utils.cpp]: Fixed. https://github.com/intel/intel-extension-for-transformers/pull/376/commits/f229a212e552141ddea739ed47ef6b7133931e68

File: intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h (hunk 3 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/baichuan/baichuan.h]: Fixed. https://github.com/intel/intel-extension-for-transformers/pull/376/commits/f229a212e552141ddea739ed47ef6b7133931e68

**Review by a32543254 — APPROVED** [review]
LGTM

**Pull request merged**